File: F# Workspace/02_HigherOrder/02_HigherOrder/combine_three.fs
(*
 * combine_three.fs
 * By Antonio F. Huertas
 * This program displays the the result of combining three integers.
 *)

// sum : int -> int -> int
// Returns the sum of two integers.
let sum val1 val2 =  val1 + val2

// product : int -> int -> int
// Returns the product of two integers.
let product val1 val2 =  val1 * val2

// minimum : 'a -> 'a -> 'a when 'a : comparison
// Returns the minimum of two values.
let minimum val1 val2 =
    if val1 < val2 then val1 else val2

// combine_three : 'a -> 'b -> 'b -> ('a -> 'b -> 'a) -> 'a
// Returns the result of combining three values.
let combine_three val1 val2 val3 combine_two =
    let result = combine_two val1 val2
    combine_two result val3

// start : int -> int -> int -> unit
// Serves as the entry-point of the program.
let start val1 val2 val3 =
    printfn "The numbers are %d, %d and %d" val1 val2 val3
    printfn "The sum of the values is %d" 
        (combine_three val1 val2 val3 sum)
    printfn "The product of the values is %d" 
        (combine_three val1 val2 val3 product)
    printfn "The minimum value is %d" 
        (combine_three val1 val2 val3 minimum)

File: C_Projects/02_Pointers/combine_three.c

/*
 * combine_three.c
 * By Antonio F. Huertas
 * This program displays the the result of combining three integers
 * using different methods.
 */

#include <stdio.h>
// for printf, scanf

// Declares functions using prototypes.
int sum(int val1, int val2);
int product(int val1, int val2);
int minimum(int val1, int val2);
int combine_three(int val1, int val2, int val3,
                  int (*combine_two)(int val1, int val2));

// Starts the execution of the program.
int main(void) {
    int value1, value2, value3;
    printf("Enter three integers: ");
    scanf("%d %d %d", &value1, &value2, &value3);

    printf("The sum of the values is %d\n",
           combine_three(value1, value2, value3, &sum));
    printf("The product of the values is %d\n",
           combine_three(value1, value2, value3, &product));
    printf("The minimum value is %d\n",
           combine_three(value1, value2, value3, &minimum));

    return 0;
}

// Returns the sum of two integers.
int sum(int val1, int val2) {
    return val1 + val2;
}

// Returns the product of two integers.
int product(int val1, int val2) {
    return val1 * val2;
}

// Returns the minimum of two integers.
int minimum(int val1, int val2) {
    return (val1 < val2)? val1 : val2;
}

// Returns the result of combining three integers.
int combine_three(int val1, int val2, int val3,
                  int (*combine_two)(int val1, int val2)) {
    int result = (*combine_two)(val1, val2);
    return (*combine_two)(result, val3);
}
